pkg/credentialprovider/azure_credentials.go

```diff
@@ -55,45 +61,73 @@ type CredentialProvider interface {
 
 // acrProvider implements the credential provider interface for Azure Container Registry.
 type acrProvider struct {
-	config                *providerconfig.AzureAuthConfig
-	environment           *azure.Environment
-	servicePrincipalToken *adal.ServicePrincipalToken
+	config      *providerconfig.AzureAuthConfig
+	environment *azure.Environment
+	credential  azcore.TokenCredential
+}
+
+func NewAcrProvider(config *providerconfig.AzureAuthConfig, environment *azure.Environment, credential azcore.TokenCredential) CredentialProvider {
+	return &acrProvider{
+		config:      config,
+		credential:  credential,
+		environment: environment,
+	}
 }
 
 // NewAcrProvider creates a new instance of the ACR provider.
-func NewAcrProvider(configFile string) (CredentialProvider, error) {
+func NewAcrProviderFromConfig(configFile string) (CredentialProvider, error) {
 	if len(configFile) == 0 {
 		return nil, errors.New("no azure credential file is provided")
 	}
-
-	f, err := os.Open(configFile)
+	config, err := configloader.Load[providerconfig.AzureAuthConfig](context.Background(), nil, &configloader.FileLoaderConfig{FilePath: configFile})
 	if err != nil {
-		return nil, fmt.Errorf("failed to load config from file %s: %w", configFile, err)
+		return nil, fmt.Errorf("failed to load config: %w", err)
 	}
-	defer f.Close()
-
-	return newAcrProviderFromConfigReader(f)
-}
 
-func newAcrProviderFromConfigReader(configReader io.Reader) (*acrProvider, error) {
-	config, env, err := providerconfig.ParseAzureAuthConfig(configReader)
-	if err != nil {
-		return nil, fmt.Errorf("failed to load config: %w", err)
+	var envConfig azure.Environment
+	envFilePath, ok := os.LookupEnv(azclient.EnvironmentFilepathName)
+	if ok {
+		content, err := os.ReadFile(envFilePath)
+		if err != nil {
+			return nil, err
+		}
+		if err = json.Unmarshal(content, &envConfig); err != nil {
+			return nil, err
+		}
 	}
 
-	servicePrincipalToken, err := providerconfig.GetServicePrincipalToken(config, env, env.ServiceManagementEndpoint)
+	var managedIdentityCredential azcore.TokenCredential
+
+	clientOption, err := azclient.GetAzCoreClientOption(&config.ARMClientConfig)
 	if err != nil {
-		return nil, fmt.Errorf("failed to create service principal token: %w", err)
+		return nil, err
+	}
+	if config.UseManagedIdentityExtension {
+		credOptions := &azidentity.ManagedIdentityCredentialOptions{
+			ClientOptions: *clientOption,
+		}
+		if len(config.UserAssignedIdentityID) > 0 {
+			if strings.Contains(strings.ToUpper(config.UserAssignedIdentityID), "/SUBSCRIPTIONS/") {
+				credOptions.ID = azidentity.ResourceID(config.UserAssignedIdentityID)
+			} else {
+				credOptions.ID = azidentity.ClientID(config.UserAssignedIdentityID)
+			}
+		}
+		managedIdentityCredential, err = azidentity.NewManagedIdentityCredential(credOptions)
+		if err != nil {
+			return nil, err
+		}
+		managedIdentityCredential = armauth.NewExpireEarlyTokenWrapper(managedIdentityCredential)
```

Comment:
What about using NewAuthProvider(xx).GetAzIdentity() which wraps the these operations? for example,
```
		var authProvider *azclient.AuthProvider
		authProvider, err = azclient.NewAuthProvider(&azclients.azureConfig.ARMClientConfig, &azclients.azureConfig.AzureAuthConfig)
		if err != nil {
			return nil, err
		}
		cred := authProvider.GetAzIdentity()
```

Reply:
the env file will be parsed twice and env object is referenced in code. So, I just put essential code here.

Reply:
I think it's ok to load env file twice in short clear code, compared to maintaining the these lengthy code with multiple if/else. Please reconsider this.